[FIX] web: take arrow size into account when displaying popup

huca-odoo opened the pull request:
Currently, the arrow of a popup is not taken into account when computing its position resulting in an overlap with its parent element in some cases. This commit fixes this situation.

Steps
=====
- Install module project_enterprise
- Create a project with a milestone set at the end of the current month
- Assigne this milestone to a task with planned dates set
- Open the Gantt view of the project
- Hover the milestone diamond

Issue
=====
- The popup displaying the milestone name is partually overlapping the diamond resulting in a flicker when the mouse is positionned on the left of the diamond.

Cause
=====
The size of the arrow linked to the popup is not taken into account when computing its position. If the popup is displayed left or at the top of its source element, the size of the arrow must be withdrawn from its positon.

Fix
===
The size of the arrow, if it is displayed is removed from the current positionning.

task-3457106

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/131048).

auon-odoo commented on addons/web/static/src/core/position_hook.js:
```diff
@@ -122,17 +122,18 @@ function getBestPosition(reference, popper, { container, margin, position }) {
     /** @type {DirectionsData} */
     const directionsData = {
-        t: refBox.top - popMargins.bottom - margin - popBox.height,
+        t: refBox.top - popMargins.bottom - margin - popBox.height - arrowBox.height,
```
```suggestion
        t: refBox.top - popMargins.bottom - margin - popBox.height - arrowBox?.height ?? 0,
```

auon-odoo reviewed (changes requested):
Sounds a little too artificial to me to create an object with a height.
Instead, I suggest leaving it undefined and using a nullish coalescing operator. :slightly_smiling_face: 
Thanks for your work!

auon-odoo commented on addons/web/static/src/core/position_hook.js:
```diff
@@ -122,17 +122,18 @@ function getBestPosition(reference, popper, { container, margin, position }) {
     // Boxes
     const popBox = popper.getBoundingClientRect();
     const refBox = reference.getBoundingClientRect();
+    const arrowBox = popper.querySelector(".popover-arrow") ? popper.querySelector(".popover-arrow")?.getBoundingClientRect();
```
```suggestion
    const arrowBox = popper.querySelector(".popover-arrow")?.getBoundingClientRect();
```
:angel:

huca-odoo commented:
> Sounds a little too artificial to me to create an object with a height. Instead, I suggest leaving it undefined and using a nullish coalescing operator. slightly_smiling_face Thanks for your work!

Thank you for the review :slightly_smiling_face: I've taken your comments into account and also changed a bit the code. I realize that it is probably better to do the change in the Popover class instead of directly in the position hook. Plus I've added some tests. Let me know if you have other comments :wink:

auon-odoo commented on addons/web/static/src/core/popover/popover.js:
```diff
@@ -11,7 +11,19 @@ export class Popover extends Component {
+        
+        // Modify positioning if the popup has an arrow
+        const arrowBox = arrowEl.style.display != "none" && arrowEl.getBoundingClientRect();
+        if (arrowBox && direction === "left") {
+            left -= arrowBox.height;
+            el.style.left = `${left}px`;
+        } else if (arrowBox && direction === "top") {
+            top -= arrowBox.height;
+            el.style.top = `${top}px`;
+        }
```
```suggestion
        if (arrowBox) {
            if (direction === "left") {
                left -= arrowBox.height;
                el.style.left = `${left}px`;
            } else if (direction === "top") {
                top -= arrowBox.height;
                el.style.top = `${top}px`;
            }
        }
```
just a little rework :slightly_smiling_face:

auon-odoo replied:
Out of curiosity, why isn't needed for right and bottom?

auon-odoo commented on addons/web/static/tests/core/popover/popover_tests.js:
```diff
@@ -158,3 +158,72 @@ QUnit.test("reposition popover should properly change classNames", async (assert
+    // Force some style, to make this test independent of screen size
+    const container = document.createElement("div");
+    container.id = "container";
+    container.style.backgroundColor = "pink";
+    container.style.height = "450px";
+    container.style.width = "450px";
+    container.style.display = "flex";
+    container.style.alignItems = "center";
+    container.style.justifyContent = "center";
```
```suggestion
    container.style = {
        ...container.style,
        backgroundColor: "pink",
        etc.
    };
```
Isn't that possible? :thinking:

auon-odoo replied:
Also, is the background color necessary?

huca-odoo replied:
> Also, is the background color necessary?

Isn't it Barbie's month :unicorn: :sparkles:  ? 
Actually I've copy paste the style from the previous test which was made to avoid repositioning and it was working well so I did not change it afterward. I'll remove what can be removed :slightly_smiling_face: (I think it was initially used to be able to identify blocks in test debug mode)

auon-odoo commented on addons/web/static/tests/core/popover/popover_tests.js:
```diff
@@ -158,3 +158,72 @@ QUnit.test("reposition popover should properly change classNames", async (assert
+
+    const TestPopover = class extends Popover {
+        onPositioned(el, position) {
+            super.onPositioned(el, position);
```
Maybe an "assertStep" is needed, in order to be sure that the test doesn't pass just because onPositioned is never called? Maybe overkill :shrug:

auon-odoo reviewed (changes requested):
Thanks for the update and the test. :fire:  Indeed, Popover seems the right place to have this code! I left some comment :slightly_smiling_face:

auon-odoo reviewed (approved):
:fire: :fire:

MissingNoShiny reviewed (approved):
LGTMn thanks :slightly_smiling_face:

brboi commented on addons/web/static/src/core/popover/popover.js:
```diff
@@ -11,7 +11,21 @@ export class Popover extends Component {
         });
     }
-    onPositioned(el, { direction, variant }) {
+    onPositioned(el, { top, left, direction, variant }) {
+        const arrowEl = el.firstElementChild;
+        
+        // Modify positioning if the popup has an arrow
+        const arrowBox = arrowEl.style.display != "none" && arrowEl.getBoundingClientRect();
+        if (arrowBox) {
+            if (direction === "left") {
+                left -= arrowBox.height;
+                el.style.left = `${left}px`;
+            } else if (direction === "top") {
+                top -= arrowBox.height;
+                el.style.top = `${top}px`;
+            }
+        }
+        
```
```suggestion
```

Instead do in popover.scss
```patch
diff --git a/addons/web/static/src/core/popover/popover.scss b/addons/web/static/src/core/popover/popover.scss
index e9b38e8408b2..8bf161844648 100644
--- a/addons/web/static/src/core/popover/popover.scss
+++ b/addons/web/static/src/core/popover/popover.scss
@@ -36,7 +36,7 @@
 	}
 
 	&.o-popover-top, &.o-popover-auto[x-placement^="top"] {
-		margin-bottom: 0.5rem;
+		margin-top: -0.5rem;
 	}
 
 	&.o-popover-right, &.o-popover-auto[x-placement^="right"] {
@@ -48,7 +48,7 @@
 	}
 
 	&.o-popover-left, &.o-popover-auto[x-placement^="left"] {
-		margin-right: 0.5rem;
+		margin-left: -0.5rem;
 	}
 
 	&.o-popover--ts .popover-arrow, &.o-popover--bs .popover-arrow {
```

huca-odoo replied:
> In fact I have a better fix:
> 
> * replace
> 
> ```css
> 	&.o-popover-top, &.o-popover-auto[x-placement^="top"] {
> 		margin-bottom: 0.5rem;
> 	}
> 
> 	&.o-popover-right, &.o-popover-auto[x-placement^="right"] {
> 		margin-left: 0.5rem;
> 	}
> 
> 	&.o-popover-bottom, &.o-popover-auto[x-placement^="bottom"] {
> 		margin-top: 0.5rem;
> 	}
> 
> 	&.o-popover-left, &.o-popover-auto[x-placement^="left"] {
> 		margin-right: 0.5rem;
> 	}
> ```
> 
> * with
> 
> ```css
> 	margin: 0.5rem;
> ```
> 
> I gave it a try in 16.0 and in master branches.
> 
> You'll need to make it conditional in master though, as the arrow can now be disabled.

Thank you @brboi for the feedback and sorry for the delay to make the changes. Your last suggestion mess up with the arrow when it is displayed (e.g. for the activity popup in saas-16.3, see the task description here: https://www.odoo.com/web#id=3457106&cids=1&menu_id=4720&action=333&active_id=4105&model=project.task&view_type=form)

The fact that we add left margin when it is not needed (when the popup is positioned above/below an element) mess the alignment of the arrow. So What I've done here
- Remove my changes in javascript and use your penultimate suggestion
- Remove the test as the change does only affect style now
- Re-target the pr to 16.0

Let me know if you agree with that.

brboi reviewed (changes requested):
Please do this instead and retarget the PR to 16.0 branch.
The issue has originally been introduced with 23facbfb76c6e8b23137cddda5f5113273555498

xavierbol commented:
Hello @brboi ,

Thanks for your review! 🙂 

Can we have another review when you have time? 

Thanks in advance! 

Have a nice evening! 🙂

aab-odoo reviewed (approved):
robodoo r+
